File: server/internal/room/room.go
// Package room implements the pre-game lobby and its state machine (GDD §7,
// project plan §3): joining, team/element selection, bots filling empty
// slots, spectator join during in-progress matches, claim-for-rematch, and
// the transition into an authoritative game.World once a match starts.
//
// The package depends only on internal/game (to build the World at match
// start); it intentionally has no dependency on internal/bot or
// internal/ws, so bot AI and transport can be developed independently and
// wired together later at the composition root (cmd/mageserver).
package room

import (
	"fmt"

	"mage-craft/server/internal/game"
)

// State is the room's lifecycle stage.
type State string

const (
	StateLobby      State = "lobby"
	StateInProgress State = "in_progress"
	StateEnded      State = "ended"
)

const (
	minTeamSize    = 1
	maxTeamSize    = 6
	maxSpectators  = 8
	defaultBotDiff = "normal"
)

// Slot is one seat on a team: either empty, held by a connected human
// player, or filled by a bot.
type Slot struct {
	ID   string
	Team game.Team

	PlayerID string // "" if this is a bot slot
	Name     string

	IsBot      bool
	Difficulty string // only meaningful when IsBot

	Element game.ElementID // "" until chosen
	Ready   bool

	// PendingClaimPlayerID is set while a spectator has reserved this bot
	// (or empty) slot for the next rematch; the bot keeps playing until then.
	PendingClaimPlayerID string
}

// Spectator is a human watching an in-progress match.
type Spectator struct {
	PlayerID      string
	Name          string
	ClaimedSlotID string
}

type member struct {
	name string
	team *game.Team // nil until SelectTeam
}

// Room holds one match's lobby state and, once started, its live world.
type Room struct {
	ID       string
	TeamSize int

	// FillBots / BotDifficulty are set at create time; the transport layer
	// calls FillEmptyWithBots once the host has taken a seat.
	FillBots      bool
	BotDifficulty string

	state State

	members    map[string]*member // playerID -> lobby membership (seated humans)
	spectators map[string]*Spectator
	slots      map[game.Team][]*Slot
	nextSeq    int

	world *game.World
}

// NewRoom creates a room with team size between 1 and 6 (capacity =
// 2*teamSize, GDD §7).
func NewRoom(id string, teamSize int) (*Room, error) {
	if teamSize < minTeamSize || teamSize > maxTeamSize {
		return nil, fmt.Errorf("room: teamSize must be between %d and %d, got %d", minTeamSize, maxTeamSize, teamSize)
	}
	return &Room{
		ID:            id,
		TeamSize:      teamSize,
		BotDifficulty: defaultBotDiff,
		state:         StateLobby,
		members:       make(map[string]*member),
		spectators:    make(map[string]*Spectator),
		slots: map[game.Team][]*Slot{
			game.TeamA: {},
			game.TeamB: {},
		},
	}, nil
}

// State reports the room's current lifecycle stage.
func (r *Room) State() State { return r.state }

// Join registers a connected player in the room's lobby without a team yet
// (the client is expected to follow up with SelectTeam).
func (r *Room) Join(playerID, name string) error {
	if r.state != StateLobby {
		return fmt.Errorf("room: cannot join, room is %s", r.state)
	}
	if _, exists := r.members[playerID]; exists {
		return fmt.Errorf("room: player %q already joined", playerID)
	}
	if _, exists := r.spectators[playerID]; exists {
		return fmt.Errorf("room: player %q already spectating", playerID)
	}
	if r.occupiedCount() >= r.capacity() {
		return fmt.Errorf("room: room is full")
	}
	r.members[playerID] = &member{name: name}
	return nil
}

// JoinAsSpectator registers a viewer during an in-progress match. They do
// not occupy a team slot until they claim one and rematch applies the claim.
func (r *Room) JoinAsSpectator(playerID, name string) error {
	if r.state != StateInProgress {
		return fmt.Errorf("room: cannot spectate, room is %s", r.state)
	}
	if _, exists := r.members[playerID]; exists {
		return fmt.Errorf("room: player %q already joined", playerID)
	}
	if _, exists := r.spectators[playerID]; exists {
		return fmt.Errorf("room: player %q already spectating", playerID)
	}
	if len(r.spectators) >= maxSpectators {
		return fmt.Errorf("room: spectator limit reached")
	}
	r.spectators[playerID] = &Spectator{PlayerID: playerID, Name: name}
	return nil
}

// Leave removes a player from the room (member or spectator), freeing their
// slot / pending claim.
func (r *Room) Leave(playerID string) error {
	if spec, ok := r.spectators[playerID]; ok {
		if spec.ClaimedSlotID != "" {
			r.clearClaimOnSlot(spec.ClaimedSlotID)
		}
		delete(r.spectators, playerID)
		return nil
	}
	mem, ok := r.members[playerID]
	if !ok {
		return fmt.Errorf("room: player %q is not in this room", playerID)
	}
	if mem.team != nil {
		r.removeSlotFor(*mem.team, playerID)
	}
	delete(r.members, playerID)
	return nil
}

// SelectTeam assigns (or re-assigns) a player to team 0 or 1, allocating
// them a slot. It fails if that team's slots are already full.
func (r *Room) SelectTeam(playerID string, team game.Team) error {
	if r.state != StateLobby {
		return fmt.Errorf("room: cannot select team, room is %s", r.state)
	}
	mem, ok := r.members[playerID]
	if !ok {
		return fmt.Errorf("room: player %q has not joined this room", playerID)
	}
	if mem.team != nil && *mem.team == team {
		return nil
	}
	if r.humanBotCountForTeam(team) >= r.TeamSize {
		return fmt.Errorf("room: team %d is full", team)
	}

	if mem.team != nil {
		r.removeSlotFor(*mem.team, playerID)
	}

	r.nextSeq++
	slot := &Slot{
		ID:       fmt.Sprintf("%s-slot-%d", r.ID, r.nextSeq),
		Team:     team,
		PlayerID: playerID,
		Name:     mem.name,
	}
	r.slots[team] = append(r.slots[team], slot)
	mem.team = &team
	return nil
}

// CanSelectElement reports whether element is still free on the given team
// (GDD §7 uniqueness rule: no two mages on the same team share an element).
func (r *Room) CanSelectElement(team game.Team, element game.ElementID) bool {
	for _, s := range r.slots[team] {
		if s.Element == element {
			return false
		}
	}
	return true
}

// SelectElement sets the caller's element, validating that it exists in the
// catalog and is still free on their team.
func (r *Room) SelectElement(playerID string, element game.ElementID) error {
	if r.state != StateLobby {
		return fmt.Errorf("room: cannot select element, room is %s", r.state)
	}
	mem, ok := r.members[playerID]
	if !ok {
		return fmt.Errorf("room: player %q has not joined this room", playerID)
	}
	if mem.team == nil {
		return fmt.Errorf("room: player %q must select a team before an element", playerID)
	}
	if _, ok := game.ElementDefFor(element); !ok {
		return fmt.Errorf("room: unknown element %q", element)
	}
	if !r.CanSelectElement(*mem.team, element) {
		return fmt.Errorf("room: element %q is already taken on team %d", element, *mem.team)
	}

	slot := r.findSlot(*mem.team, playerID)
	if slot == nil {
		return fmt.Errorf("room: internal error: no slot found for player %q", playerID)
	}
	slot.Element = element
	return nil
}

// AddBot fills an empty seat on team with a bot, auto-selecting the first
// element from the catalog that isn't already used on that team (GDD §7).
func (r *Room) AddBot(team game.Team, difficulty string) (*Slot, error) {
	if r.state != StateLobby {
		return nil, fmt.Errorf("room: cannot add bot, room is %s", r.state)
	}
	if r.humanBotCountForTeam(team) >= r.TeamSize {
		return nil, fmt.Errorf("room: team %d is full", team)
	}
	if difficulty == "" {
		difficulty = defaultBotDiff
	}

	element, ok := r.firstFreeElement(team)
	if !ok {
		return nil, fmt.Errorf("room: no free element left on team %d", team)
	}

	r.nextSeq++
	slot := &Slot{
		ID:         fmt.Sprintf("%s-slot-%d", r.ID, r.nextSeq),
		Team:       team,
		IsBot:      true,
		Difficulty: difficulty,
		Element:    element,
		Ready:      true,
		Name:       fmt.Sprintf("Bot (%s)", difficulty),
	}
	r.slots[team] = append(r.slots[team], slot)
	return slot, nil
}

// FillEmptyWithBots fills every remaining seat on both teams with bots.
func (r *Room) FillEmptyWithBots(difficulty string) error {
	if r.state != StateLobby {
		return fmt.Errorf("room: cannot fill bots, room is %s", r.state)
	}
	if difficulty == "" {
		difficulty = r.BotDifficulty
	}
	if difficulty == "" {
		difficulty = defaultBotDiff
	}
	for _, team := range []game.Team{game.TeamA, game.TeamB} {
		for r.humanBotCountForTeam(team) < r.TeamSize {
			if _, err := r.AddBot(team, difficulty); err != nil {
				return err
			}
		}
	}
	return nil
}

// RemoveBot removes a previously added bot slot, freeing its team capacity
// and element. Pending claims on that slot are cleared.
func (r *Room) RemoveBot(slotID string) error {
	if r.state != StateLobby {
		return fmt.Errorf("room: cannot remove bot, room is %s", r.state)
	}
	for team, slots := range r.slots {
		for i, s := range slots {
			if s.ID == slotID && s.IsBot {
				if s.PendingClaimPlayerID != "" {
					if spec := r.spectators[s.PendingClaimPlayerID]; spec != nil {
						spec.ClaimedSlotID = ""
					}
				}
				r.slots[team] = append(slots[:i], slots[i+1:]...)
				return nil
			}
		}
	}
	return fmt.Errorf("room: no bot slot %q found", slotID)
}

// ClaimSlot lets a spectator reserve a bot slot for the next rematch.
// The bot keeps playing until ApplyClaims runs at round end.
func (r *Room) ClaimSlot(playerID, slotID string) error {
	if r.state != StateInProgress && r.state != StateLobby {
		return fmt.Errorf("room: cannot claim slot, room is %s", r.state)
	}
	spec, ok := r.spectators[playerID]
	if !ok {
		return fmt.Errorf("room: player %q is not a spectator", playerID)
	}
	slot := r.findSlotByID(slotID)
	if slot == nil {
		return fmt.Errorf("room: slot %q not found", slotID)
	}
	if !slot.IsBot && slot.PlayerID != "" {
		return fmt.Errorf("room: slot %q is occupied by a human", slotID)
	}
	if slot.PendingClaimPlayerID != "" && slot.PendingClaimPlayerID != playerID {
		return fmt.Errorf("room: slot %q already claimed", slotID)
	}
	if spec.ClaimedSlotID != "" && spec.ClaimedSlotID != slotID {
		r.clearClaimOnSlot(spec.ClaimedSlotID)
	}
	slot.PendingClaimPlayerID = playerID
	spec.ClaimedSlotID = slotID
	return nil
}

// SetReady toggles a connected player's ready flag.
func (r *Room) SetReady(playerID string, ready bool) error {
	mem, ok := r.members[playerID]
	if !ok {
		return fmt.Errorf("room: player %q has not joined this room", playerID)
	}
	if mem.team == nil {
		return fmt.Errorf("room: player %q must select a team first", playerID)
	}
	slot := r.findSlot(*mem.team, playerID)
	if slot == nil {
		return fmt.Errorf("room: internal error: no slot found for player %q", playerID)
	}
	slot.Ready = ready
	return nil
}

// StartMatch validates that both teams are completely filled with valid
// elements, builds the authoritative game.World for the match, and
// transitions the room into StateInProgress.
func (r *Room) StartMatch() (*game.World, error) {
	if r.state != StateLobby {
		return nil, fmt.Errorf("room: cannot start match, room is %s", r.state)
	}
	for _, team := range []game.Team{game.TeamA, game.TeamB} {
		slots := r.slots[team]
		if len(slots) != r.TeamSize {
			return nil, fmt.Errorf("room: team %d has %d/%d slots filled", team, len(slots), r.TeamSize)
		}
		for _, s := range slots {
			if s.Element == "" {
				return nil, fmt.Errorf("room: slot %q on team %d has no element selected", s.ID, team)
			}
		}
	}

	world := game.NewWorld()
	for _, team := range []game.Team{game.TeamA, game.TeamB} {
		for _, s := range r.slots[team] {
			id := s.PlayerID
			if s.IsBot {
				id = s.ID
			}
			world.AddMage(id, team, s.Element, s.IsBot)
		}
	}

	r.world = world
	r.state = StateInProgress
	return world, nil
}

// World returns the room's live simulation once a match has started, or nil
// before that.
func (r *Room) World() *game.World { return r.world }

// RoleOf reports "player", "spectator", or "" if the id is unknown.
func (r *Room) RoleOf(playerID string) string {
	if _, ok := r.members[playerID]; ok {
		return "player"
	}
	if _, ok := r.spectators[playerID]; ok {
		return "spectator"
	}
	return ""
}

// MemberIDs returns every human connected to the room (seated players and
// spectators), so the transport can broadcast room_state / snapshots to all.
func (r *Room) MemberIDs() []string {
	out := make([]string, 0, len(r.members)+len(r.spectators))
	for id := range r.members {
		out = append(out, id)
	}
	for id := range r.spectators {
		out = append(out, id)
	}
	return out
}

// Spectators returns a snapshot of current spectators.
func (r *Room) Spectators() []Spectator {
	out := make([]Spectator, 0, len(r.spectators))
	for _, s := range r.spectators {
		out = append(out, *s)
	}
	return out
}

// Slots returns a value-copy snapshot of every occupied slot across both
// teams.
func (r *Room) Slots() []Slot {
	out := make([]Slot, 0, r.TeamSize*2)
	for _, team := range []game.Team{game.TeamA, game.TeamB} {
		for _, s := range r.slots[team] {
			out = append(out, *s)
		}
	}
	return out
}

// Summary is a compact view used by list_rooms.
type Summary struct {
	RoomID            string
	TeamSize          int
	State             State
	Filled            int
	Capacity          int
	OpenBotSlots      int
	AcceptsSpectators bool
}

// Summary builds the list_rooms entry for this room.
func (r *Room) Summary() Summary {
	filled := 0
	openBots := 0
	for _, s := range r.Slots() {
		filled++
		if s.IsBot && s.PendingClaimPlayerID == "" {
			openBots++
		}
	}
	return Summary{
		RoomID:            r.ID,
		TeamSize:          r.TeamSize,
		State:             r.state,
		Filled:            filled,
		Capacity:          r.capacity(),
		OpenBotSlots:      openBots,
		AcceptsSpectators: r.state == StateInProgress && len(r.spectators) < maxSpectators,
	}
}

// MarkEnded transitions the room to StateEnded (abandon path). Rematch uses
// ResetToLobby instead.
func (r *Room) MarkEnded() {
	r.state = StateEnded
}

// ApplyClaims converts pending spectator claims into seated human slots,
// then clears the spectator roster entries that were promoted.
func (r *Room) ApplyClaims() {
	for _, team := range []game.Team{game.TeamA, game.TeamB} {
		for _, s := range r.slots[team] {
			if s.PendingClaimPlayerID == "" {
				continue
			}
			spec, ok := r.spectators[s.PendingClaimPlayerID]
			if !ok {
				s.PendingClaimPlayerID = ""
				continue
			}
			s.IsBot = false
			s.Difficulty = ""
			s.PlayerID = spec.PlayerID
			s.Name = spec.Name
			s.Ready = false
			s.PendingClaimPlayerID = ""
			teamCopy := team
			r.members[spec.PlayerID] = &member{name: spec.Name, team: &teamCopy}
			delete(r.spectators, spec.PlayerID)
		}
	}
}

// ResetToLobby clears the live world and returns the room to lobby for rematch.
// Call ApplyClaims before or after; Session.BeginRematch does ApplyClaims first.
func (r *Room) ResetToLobby() {
	r.world = nil
	r.state = StateLobby
	// Humans keep their seats; bots stay until claimed/removed. Clear ready
	// so everyone confirms for the next round.
	for _, team := range []game.Team{game.TeamA, game.TeamB} {
		for _, s := range r.slots[team] {
			if !s.IsBot {
				s.Ready = false
			}
		}
	}
}

func (r *Room) capacity() int { return r.TeamSize * 2 }

func (r *Room) occupiedCount() int {
	return len(r.members) + len(r.slots[game.TeamA]) + len(r.slots[game.TeamB]) - r.humanMembersWithTeam()
}

// humanMembersWithTeam avoids double counting: members that already have a
// team are represented both in r.members and in a team's slot list.
func (r *Room) humanMembersWithTeam() int {
	n := 0
	for _, m := range r.members {
		if m.team != nil {
			n++
		}
	}
	return n
}

func (r *Room) humanBotCountForTeam(team game.Team) int {
	return len(r.slots[team])
}

func (r *Room) firstFreeElement(team game.Team) (game.ElementID, bool) {
	for _, id := range game.AllElements() {
		if r.CanSelectElement(team, id) {
			return id, true
		}
	}
	return "", false
}

func (r *Room) findSlot(team game.Team, playerID string) *Slot {
	for _, s := range r.slots[team] {
		if s.PlayerID == playerID {
			return s
		}
	}
	return nil
}

func (r *Room) findSlotByID(slotID string) *Slot {
	for _, team := range []game.Team{game.TeamA, game.TeamB} {
		for _, s := range r.slots[team] {
			if s.ID == slotID {
				return s
			}
		}
	}
	return nil
}

func (r *Room) clearClaimOnSlot(slotID string) {
	if s := r.findSlotByID(slotID); s != nil {
		s.PendingClaimPlayerID = ""
	}
}

func (r *Room) removeSlotFor(team game.Team, playerID string) {
	slots := r.slots[team]
	for i, s := range slots {
		if s.PlayerID == playerID {
			r.slots[team] = append(slots[:i], slots[i+1:]...)
			return
		}
	}
}

File: server/src/Room.ts
/**
 * The pre-game lobby and its state machine (GDD §7): joining, team/element
 * selection, bots filling empty slots, spectator join during in-progress
 * matches, claim-for-rematch, and the transition into an authoritative
 * simulation once a match starts.
 *
 * Depends only on the shared `sim` package (to build the World at match start);
 * it deliberately knows nothing about bot AI or the WebSocket transport, so
 * those stay independently testable and are wired together in App.ts.
 */

import { ALL_ELEMENTS, elementDefFor, type ElementId } from '../../sim/elements';
import { TEAM_A, TEAM_B, type Team } from '../../sim/entities';
import { World } from '../../sim/World';

export type RoomState = 'lobby' | 'in_progress' | 'ended';

export const MIN_TEAM_SIZE = 1;
export const MAX_TEAM_SIZE = 6;
const MAX_SPECTATORS = 8;
const DEFAULT_BOT_DIFFICULTY = 'normal';

const TEAMS: readonly Team[] = [TEAM_A, TEAM_B];

/** One seat on a team: empty, held by a connected human, or filled by a bot. */
export interface Slot {
  id: string;
  team: Team;

  /** Empty for a bot slot. */
  playerId: string;
  name: string;

  isBot: boolean;
  /** Only meaningful when `isBot`. */
  difficulty: string;

  /** Null until chosen. */
  element: ElementId | null;
  ready: boolean;

  /**
   * Set while a spectator has reserved this bot (or empty) slot for the next
   * rematch; the bot keeps playing until then.
   */
  pendingClaimPlayerId: string;
}

/** A human watching an in-progress match. */
export interface Spectator {
  playerId: string;
  name: string;
  claimedSlotId: string;
}

interface Member {
  name: string;
  /** Null until selectTeam. */
  team: Team | null;
}

/** The compact view used by `list_rooms`. */
export interface RoomSummary {
  roomId: string;
  teamSize: number;
  state: RoomState;
  filled: number;
  capacity: number;
  openBotSlots: number;
  acceptsSpectators: boolean;
}

export class Room {
  /** Set at create time; App calls fillEmptyWithBots once the host has a seat. */
  fillBots = false;
  botDifficulty = DEFAULT_BOT_DIFFICULTY;

  private roomState: RoomState = 'lobby';

  private readonly members = new Map<string, Member>();
  private readonly spectatorsById = new Map<string, Spectator>();
  private readonly slotsByTeam = new Map<Team, Slot[]>([
    [TEAM_A, []],
    [TEAM_B, []],
  ]);
  private nextSeq = 0;

  private liveWorld: World | null = null;

  /** Team size is between 1 and 6; capacity is 2 * teamSize (GDD §7). */
  constructor(
    readonly id: string,
    readonly teamSize: number,
  ) {
    if (!Number.isInteger(teamSize) || teamSize < MIN_TEAM_SIZE || teamSize > MAX_TEAM_SIZE) {
      throw new Error(
        `room: teamSize must be an integer between ${MIN_TEAM_SIZE} and ${MAX_TEAM_SIZE}, got ${teamSize}`,
      );
    }
  }

  get state(): RoomState {
    return this.roomState;
  }

  /** The live simulation once a match has started, or null before that. */
  get world(): World | null {
    return this.liveWorld;
  }

  /** Registers a connected player in the lobby without a team yet. */
  join(playerId: string, name: string): void {
    if (this.roomState !== 'lobby') {
      throw new Error(`room: cannot join, room is ${this.roomState}`);
    }
    if (this.members.has(playerId)) throw new Error(`room: player ${playerId} already joined`);
    if (this.spectatorsById.has(playerId)) {
      throw new Error(`room: player ${playerId} already spectating`);
    }
    if (this.occupiedCount() >= this.capacity()) throw new Error('room: room is full');

    this.members.set(playerId, { name, team: null });
  }

  /**
   * Registers a viewer during an in-progress match. They do not occupy a team
   * slot until they claim one and a rematch applies the claim.
   */
  joinAsSpectator(playerId: string, name: string): void {
    if (this.roomState !== 'in_progress') {
      throw new Error(`room: cannot spectate, room is ${this.roomState}`);
    }
    if (this.members.has(playerId)) throw new Error(`room: player ${playerId} already joined`);
    if (this.spectatorsById.has(playerId)) {
      throw new Error(`room: player ${playerId} already spectating`);
    }
    if (this.spectatorsById.size >= MAX_SPECTATORS) throw new Error('room: spectator limit reached');

    this.spectatorsById.set(playerId, { playerId, name, claimedSlotId: '' });
  }

  /** Removes a player (member or spectator), freeing their slot / pending claim. */
  leave(playerId: string): void {
    const spec = this.spectatorsById.get(playerId);
    if (spec) {
      if (spec.claimedSlotId) this.clearClaimOnSlot(spec.claimedSlotId);
      this.spectatorsById.delete(playerId);
      return;
    }

    const mem = this.members.get(playerId);
    if (!mem) throw new Error(`room: player ${playerId} is not in this room`);
    if (mem.team !== null) this.removeSlotFor(mem.team, playerId);
    this.members.delete(playerId);
  }

  /** Assigns (or re-assigns) a player to a team, allocating them a slot. */
  selectTeam(playerId: string, team: Team): void {
    if (this.roomState !== 'lobby') {
      throw new Error(`room: cannot select team, room is ${this.roomState}`);
    }
    const mem = this.members.get(playerId);
    if (!mem) throw new Error(`room: player ${playerId} has not joined this room`);
    if (mem.team === team) return;
    if (this.slotsFor(team).length >= this.teamSize) throw new Error(`room: team ${team} is full`);

    if (mem.team !== null) this.removeSlotFor(mem.team, playerId);

    this.nextSeq++;
    this.slotsFor(team).push({
      id: `${this.id}-slot-${this.nextSeq}`,
      team,
      playerId,
      name: mem.name,
      isBot: false,
      difficulty: '',
      element: null,
      ready: false,
      pendingClaimPlayerId: '',
    });
    mem.team = team;
  }

  /**
   * Whether `element` is still free on a team (GDD §7 uniqueness rule: no two
   * mages on the same team share an element).
   */
  canSelectElement(team: Team, element: ElementId): boolean {
    return !this.slotsFor(team).some((s) => s.element === element);
  }

  selectElement(playerId: string, element: string): void {
    if (this.roomState !== 'lobby') {
      throw new Error(`room: cannot select element, room is ${this.roomState}`);
    }
    const mem = this.members.get(playerId);
    if (!mem) throw new Error(`room: player ${playerId} has not joined this room`);
    if (mem.team === null) {
      throw new Error(`room: player ${playerId} must select a team before an element`);
    }
    if (!elementDefFor(element as ElementId)) {
      throw new Error(`room: unknown element ${JSON.stringify(element)}`);
    }
    if (!this.canSelectElement(mem.team, element as ElementId)) {
      throw new Error(`room: element ${JSON.stringify(element)} is already taken on team ${mem.team}`);
    }

    const slot = this.findSlot(mem.team, playerId);
    if (!slot) throw new Error(`room: internal error: no slot found for player ${playerId}`);
    slot.element = element as ElementId;
  }

  /**
   * Fills an empty seat with a bot, auto-selecting the first catalog element
   * that isn't already used on that team (GDD §7).
   */
  addBot(team: Team, difficulty: string): Slot {
    if (this.roomState !== 'lobby') {
      throw new Error(`room: cannot add bot, room is ${this.roomState}`);
    }
    if (this.slotsFor(team).length >= this.teamSize) throw new Error(`room: team ${team} is full`);

    const diff = difficulty || DEFAULT_BOT_DIFFICULTY;
    const element = this.firstFreeElement(team);
    if (!element) throw new Error(`room: no free element left on team ${team}`);

    this.nextSeq++;
    const slot: Slot = {
      id: `${this.id}-slot-${this.nextSeq}`,
      team,
      playerId: '',
      name: `Bot (${diff})`,
      isBot: true,
      difficulty: diff,
      element,
      ready: true,
      pendingClaimPlayerId: '',
    };
    this.slotsFor(team).push(slot);
    return slot;
  }

  /** Fills every remaining seat on both teams with bots. */
  fillEmptyWithBots(difficulty: string): void {
    if (this.roomState !== 'lobby') {
      throw new Error(`room: cannot fill bots, room is ${this.roomState}`);
    }
    const diff = difficulty || this.botDifficulty || DEFAULT_BOT_DIFFICULTY;
    for (const team of TEAMS) {
      while (this.slotsFor(team).length < this.teamSize) this.addBot(team, diff);
    }
  }

  /** Removes a bot slot, freeing its team capacity and element. */
  removeBot(slotId: string): void {
    if (this.roomState !== 'lobby') {
      throw new Error(`room: cannot remove bot, room is ${this.roomState}`);
    }
    for (const team of TEAMS) {
      const slots = this.slotsFor(team);
      const i = slots.findIndex((s) => s.id === slotId && s.isBot);
      if (i === -1) continue;

      const claimant = slots[i].pendingClaimPlayerId;
      if (claimant) {
        const spec = this.spectatorsById.get(claimant);
        if (spec) spec.claimedSlotId = '';
      }
      slots.splice(i, 1);
      return;
    }
    throw new Error(`room: no bot slot ${JSON.stringify(slotId)} found`);
  }

  /**
   * Lets a spectator reserve a bot slot for the next rematch. The bot keeps
   * playing until applyClaims runs at round end.
   */
  claimSlot(playerId: string, slotId: string): void {
    if (this.roomState !== 'in_progress' && this.roomState !== 'lobby') {
      throw new Error(`room: cannot claim slot, room is ${this.roomState}`);
    }
    const spec = this.spectatorsById.get(playerId);
    if (!spec) throw new Error(`room: player ${playerId} is not a spectator`);

    const slot = this.findSlotById(slotId);
    if (!slot) throw new Error(`room: slot ${JSON.stringify(slotId)} not found`);
    if (!slot.isBot && slot.playerId) {
      throw new Error(`room: slot ${JSON.stringify(slotId)} is occupied by a human`);
    }
    if (slot.pendingClaimPlayerId && slot.pendingClaimPlayerId !== playerId) {
      throw new Error(`room: slot ${JSON.stringify(slotId)} already claimed`);
    }
    if (spec.claimedSlotId && spec.claimedSlotId !== slotId) {
      this.clearClaimOnSlot(spec.claimedSlotId);
    }

    slot.pendingClaimPlayerId = playerId;
    spec.claimedSlotId = slotId;
  }

  setReady(playerId: string, ready: boolean): void {
    const mem = this.members.get(playerId);
    if (!mem) throw new Error(`room: player ${playerId} has not joined this room`);
    if (mem.team === null) throw new Error(`room: player ${playerId} must select a team first`);

    const slot = this.findSlot(mem.team, playerId);
    if (!slot) throw new Error(`room: internal error: no slot found for player ${playerId}`);
    slot.ready = ready;
  }

  /**
   * Validates that both teams are completely filled with valid elements, builds
   * the authoritative World, and moves the room to in_progress.
   */
  startMatch(): World {
    if (this.roomState !== 'lobby') {
      throw new Error(`room: cannot start match, room is ${this.roomState}`);
    }
    for (const team of TEAMS) {
      const slots = this.slotsFor(team);
      if (slots.length !== this.teamSize) {
        throw new Error(`room: team ${team} has ${slots.length}/${this.teamSize} slots filled`);
      }
      for (const s of slots) {
        if (!s.element) {
          throw new Error(
            `room: slot ${JSON.stringify(s.id)} on team ${team} has no element selected`,
          );
        }
      }
    }

    /*
     * Room only builds the empty World — the squads themselves are Session's
     * job (`World.initSquad`, GDD §4, §7), since Session is what knows the
     * bot roster and the AI commanders that squad needs wiring to. Slot and
     * element selection survive above purely as lobby state for custom rooms
     * — the simulation no longer reads either.
     */
    const world = new World();

    this.liveWorld = world;
    this.roomState = 'in_progress';
    return world;
  }

  /** "player", "spectator", or "" if the id is unknown. */
  roleOf(playerId: string): string {
    if (this.members.has(playerId)) return 'player';
    if (this.spectatorsById.has(playerId)) return 'spectator';
    return '';
  }

  /** Every human connected to the room (seated players and spectators). */
  memberIds(): string[] {
    return [...this.members.keys(), ...this.spectatorsById.keys()];
  }

  spectators(): Spectator[] {
    return [...this.spectatorsById.values()].map((s) => ({ ...s }));
  }

  /** A value-copy snapshot of every occupied slot across both teams. */
  slots(): Slot[] {
    return TEAMS.flatMap((team) => this.slotsFor(team).map((s) => ({ ...s })));
  }

  summary(): RoomSummary {
    const slots = this.slots();
    return {
      roomId: this.id,
      teamSize: this.teamSize,
      state: this.roomState,
      filled: slots.length,
      openBotSlots: slots.filter((s) => s.isBot && !s.pendingClaimPlayerId).length,
      capacity: this.capacity(),
      acceptsSpectators: this.roomState === 'in_progress' && this.spectatorsById.size < MAX_SPECTATORS,
    };
  }

  /** Abandon path; a rematch uses resetToLobby instead. */
  markEnded(): void {
    this.roomState = 'ended';
  }

  /** Converts pending spectator claims into seated human slots. */
  applyClaims(): void {
    for (const team of TEAMS) {
      for (const s of this.slotsFor(team)) {
        if (!s.pendingClaimPlayerId) continue;

        const spec = this.spectatorsById.get(s.pendingClaimPlayerId);
        if (!spec) {
          s.pendingClaimPlayerId = '';
          continue;
        }

        s.isBot = false;
        s.difficulty = '';
        s.playerId = spec.playerId;
        s.name = spec.name;
        s.ready = false;
        s.pendingClaimPlayerId = '';

        this.members.set(spec.playerId, { name: spec.name, team });
        this.spectatorsById.delete(spec.playerId);
      }
    }
  }

  /**
   * Clears the live world and returns the room to lobby for a rematch. Humans
   * keep their seats and bots stay until claimed/removed; ready is cleared so
   * everyone confirms for the next round.
   */
  resetToLobby(): void {
    this.liveWorld = null;
    this.roomState = 'lobby';
    for (const team of TEAMS) {
      for (const s of this.slotsFor(team)) {
        if (!s.isBot) s.ready = false;
      }
    }
  }

  private capacity(): number {
    return this.teamSize * 2;
  }

  private occupiedCount(): number {
    // Members that already have a team are represented both in `members` and in
    // a team's slot list, so subtract them to avoid double counting.
    const seated = [...this.members.values()].filter((m) => m.team !== null).length;
    return this.members.size + this.slotsFor(TEAM_A).length + this.slotsFor(TEAM_B).length - seated;
  }

  private slotsFor(team: Team): Slot[] {
    return this.slotsByTeam.get(team) as Slot[];
  }

  private firstFreeElement(team: Team): ElementId | null {
    return ALL_ELEMENTS.find((id) => this.canSelectElement(team, id)) ?? null;
  }

  /** Exposed for tests; App goes through slots(). */
  findSlot(team: Team, playerId: string): Slot | undefined {
    return this.slotsFor(team).find((s) => s.playerId === playerId);
  }

  private findSlotById(slotId: string): Slot | undefined {
    for (const team of TEAMS) {
      const found = this.slotsFor(team).find((s) => s.id === slotId);
      if (found) return found;
    }
    return undefined;
  }

  private clearClaimOnSlot(slotId: string): void {
    const s = this.findSlotById(slotId);
    if (s) s.pendingClaimPlayerId = '';
  }

  private removeSlotFor(team: Team, playerId: string): void {
    const slots = this.slotsFor(team);
    const i = slots.findIndex((s) => s.playerId === playerId);
    if (i !== -1) slots.splice(i, 1);
  }
}
